### ISSUE TITLE
e2e flake: Proxy version v1 [It] should proxy through a service and a pod

### ISSUE THREAD

**Issue by gmarek** [kubernetes/kubernetes#21484]
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/21251/kubernetes-pull-build-test-e2e-gce/29155/build-log.txt

```
Failure [132.205 seconds]
Proxy
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/proxy.go:40
  version v1
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/proxy.go:39
    should proxy through a service and a pod [Conformance] [It]
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/proxy.go:224

    0: path /api/v1/namespaces/e2e-tests-proxy-ehd91/pods/proxy-service-yckdo-cd23f/proxy/ gave error: an error on the server has prevented the request from succeeding

    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/proxy.go:222
```

cc @kubernetes/goog-cluster

**Comment by lavalamp** [comment]
#17493 merged recently, but it looks like after this flake happened, so it shouldn't be the cause.

**Comment by lavalamp** [comment]
Huh, somehow we returned a 0 status code.

```
Feb 18 07:48:29.875: INFO: /api/v1/namespaces/e2e-tests-proxy-ehd91/pods/proxy-service-yckdo-cd23f/proxy/: no body (0; 3.900034ms)
```

**Comment by wojtek-t** [comment]
Another failure:
http://kubekins.dls.corp.google.com:8080/view/Critical%20Builds/job/kubernetes-e2e-gce/12049/

**Comment by bprashanth** [comment]
Can we dupe this with https://github.com/kubernetes/kubernetes/issues/21484, or change the title to reflect to 0 status code issue?

**Comment by lavalamp** [comment]
@bprashanth this is #21484, did you mean something else?

**Comment by bprashanth** [comment]
Bah i meant https://github.com/kubernetes/kubernetes/issues/18963

**Comment by lavalamp** [comment]
OK, so I misread. The '0' is the index of the attempt, not the status code.

**Comment by lavalamp** [comment]
I think the pod probably just needs a readiness check.

**Comment by lavalamp** [comment]
I added one & tried it--indeed, the pod spent a fair amount of time in running but not ready state. PR sent.

**Comment by bprashanth** [comment]
https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gce/12131/?debugUI=CLOUD

**Comment by caesarxuchao** [comment]
Occurred again:
```
â€¢ Failure [154.406 seconds]
Proxy
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/proxy.go:40
  version v1
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/proxy.go:39
    should proxy through a service and a pod [Conformance] [It]
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/proxy.go:244

    0: path /api/v1/proxy/namespaces/e2e-tests-proxy-2gz9u/pods/proxy-service-j63nq-ur6c1:80/ gave error: an error on the server has prevented the request from succeeding

    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/proxy.go:242
------------------------------
```
https://console.cloud.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/21175/kubernetes-pull-build-test-e2e-gce/30727/?debugUI=CLOUD

**Comment by lavalamp** [comment]
Huh, it worked through the service but not through the pod directly.

**Comment by janetkuo** [comment]
https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/22202/kubernetes-pull-build-test-e2e-gce/31071/artifacts/?debugUI=CLOUD
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/proxy.go:244
0: path /api/v1/namespaces/e2e-tests-proxy-31hny/pods/http:proxy-service-b9v0f-v5kdh:80/proxy/ gave error: an error on the server has prevented the request from succeeding
```

**Comment by a-robinson** [comment]
http://pr-test.k8s.io/22367/kubernetes-pull-build-test-e2e-gce/31438/

**Comment by wojtek-t** [comment]
This happened again:
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce/12730/

**Comment by wojtek-t** [comment]
And one more:
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gke/2190/

**Comment by lavalamp** [comment]
@ncdc I was planning on looking at this but I think I won't get to it for the next few days---any ideas?

**Comment by ncdc** [comment]
@lavalamp not offhand, I think @liggitt has more background on how the proxying works

**Comment by liggitt** [comment]
Found this in server log from recent failed run:
https://00e9e64bacdfd420a2443c95a55e8939fcdb5dd725bc5d6585-apidata.googleusercontent.com/download/storage/v1_internal/b/kubernetes-jenkins/o/pr-logs%2Fpull%2F21964%2Fkubernetes-pull-build-test-e2e-gce%2F31403%2Fartifacts%2F104.197.156.50:22-kube-apiserver.log?qk=AD5uMEvPFqqfW4hCz72DWl9bLRwZIYQ0kQkNYmhosMlP-1A_wN4zENey37YFiLNhaDlkIiTQyNVzdQWsR4wKUwLHVza3G9D9uxFKDub0jhhLEh18WZ4DX5t26C0ZNRDYt5BNwo6-VhxUq6gx5YtqRTlRYGr0Cs66KGDXqEt2XDzRVAdq27o-zwvhYrzi1_MgtkkNr1vEndGLRmgIyM-eIvBAo-MzJ1Q5tMM5ycicKIKlZB4wQ4KW-peC6d578qLz2eprVXNcOzIPpOOAoXiax53eyJx5RD3FpuN-aL0CSlBdRDmv7r37TwIDDqE4LrO3rNeMaN0wky24KCa63lbtM6yK0JB_KBpraWqAUfmm7GmQRSGh9DgnY2hzENExhMecsH1VyFvSKQo_AU6pBLQAo1qgH5UM-q1lWVbNnLLs-nEzVSVdlFvbdDvdWGVNcOzsZmQftxVWBKt8HNGyydI6pOBwwJvuj13n6iW1KFfRYYUR_vMibgMBC4bzZ4moOHNFjkVySAnBCZK9jTU1duoeC6X2YMrcFPs--MzNXrT5DGDFcp07B4HlIVtYJ8hOq4zXC74TGQeeyWQsBmNxHojOsfe4mWE0jv1hhtyWjG9PzaAMxkBOKm77HZYREPWb3luGxr7M9owoKbXSGPCu8ewo6vjUZee3tMKh6uZrxKmCV4GvQVAkrsYpk8VrcPLWMIOmC-YyHG1421Ay-05QfOacJWfEqmEDfWflHgYkQAbHAZ88xyeIjijiGHc8kdnCsnQpx_JFLAlG9A10cJOP8OHFUOHHH9sUGBXjW5lo6kKWlzM276jyPG0stNMUoAck_FG-xyL0oc4HMLT4ydMOQmZd0zfIGdBX5gqICbH2VXzWWuRLA6O-ZAbyhEGjbEpvrsBIE11-z1eWfFjzVXO5dxocb__0pUhfEiTAD4NONst2GUITX6tDe1Dbj8OrH89uyoZqIpV8KgEQrZPhK5Y12j7HkDKuisjke_eekw

```
I0302 21:40:12.728178       6 handlers.go:152] GET /api/v1/proxy/namespaces/e2e-tests-proxy-modan/pods/proxy-service-s750o-hmka9:80/: (2.896217ms) 503
goroutine 92376 [running]:
k8s.io/kubernetes/pkg/httplog.(*respLogger).recordStatus(0xc20b389dc0, 0x1f7)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/httplog/log.go:214 +0xa6
k8s.io/kubernetes/pkg/httplog.(*respLogger).WriteHeader(0xc20b389dc0, 0x1f7)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/httplog/log.go:193 +0x32
k8s.io/kubernetes/pkg/apiserver/metrics.(*responseWriterDelegator).WriteHeader(0xc20d074c00, 0x1f7)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/metrics/metrics.go:117 +0x53
net/http/httputil.(*ReverseProxy).ServeHTTP(0xc20d075530, 0x7fb7aa714bc0, 0xc20b381d98, 0xc209cae820)
	/usr/src/go/src/net/http/httputil/reverseproxy.go:159 +0x708
k8s.io/kubernetes/pkg/apiserver.(*ProxyHandler).ServeHTTP(0xc2081e6700, 0x7fb7aa714bc0, 0xc20b381d98, 0xc209e2fe10)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/proxy.go:205 +0x2574
k8s.io/kubernetes/pkg/apiserver.funcÂ·002(0xc20d074ae0, 0xc20b26d200)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/api_installer.go:850 +0x60
k8s.io/kubernetes/pkg/apiserver/metrics.funcÂ·001(0xc20d074ae0, 0xc20b26d200)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/metrics/metrics.go:101 +0x269
github.com/emicklei/go-restful.(*Container).dispatch(0xc2082c38c0, 0x7fb7aa714868, 0xc20b389dc0, 0xc209e2fe10)
	/go/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/emicklei/go-restful/container.go:249 +0xf5e
github.com/emicklei/go-restful.*Container.(github.com/emicklei/go-restful.dispatch)Â·fm(0x7fb7aa714868, 0xc20b389dc0, 0xc209e2fe10)
	/go/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/emicklei/go-restful/container.go:95 +0x45
net/http.HandlerFunc.ServeHTTP(0xc208945a70, 0x7fb7aa714868, 0xc20b389dc0, 0xc209e2fe10)
	/usr/src/go/src/net/http/server.go:1265 +0x41
net/http.(*ServeMux).ServeHTTP(0xc208267620, 0x7fb7aa714868, 0xc20b389dc0, 0xc209e2fe10)
```

**Comment by liggitt** [comment]
And this in the e2e log:
```
Mar  2 13:40:12.729: INFO: /api/v1/proxy/namespaces/e2e-tests-proxy-modan/pods/proxy-service-s750o-hmka9:80/: no body (0; 4.560149ms)
```

**Comment by bprashanth** [comment]
The 503 is just piped back up from the kubelet/container serving right? or is this apiserver unavailable

**Comment by maisem** [comment]
Seems to have happened again 
https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gke/3092/

**Comment by fgrzadkowski** [comment]
Happened again.

```
06:31:28 • Failure [157.173 seconds]
06:31:28 [k8s.io] Proxy
06:31:28 /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework.go:420
06:31:28   version v1
06:31:28   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/proxy.go:39
06:31:28     should proxy through a service and a pod [Conformance] [It]
06:31:28     /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/proxy.go:247
06:31:28 
06:31:28     0: path /api/v1/namespaces/e2e-tests-proxy-xk6sc/pods/http:proxy-service-v0adj-rwkqf:80/proxy/ gave error: an error on the server has prevented the request from succeeding
```
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce/13685/consoleFull

**Comment by thockin** [comment]
Can I assign this to you @liggitt or someone else?  My time has been so fragmented I have not been able to get to this and I won't for at least another week or 2.

**Comment by bprashanth** [comment]
Yeah this is apiserver proxy failing to contact a pod: 
```
0: path /api/v1/proxy/namespaces/e2e-tests-proxy-hxr0y/pods/proxy-service-8l4zg-f6oym:80/ gave error: an error on the server has prevented the request from succeeding
```

I confirmed that the kublet started the pod in https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gke/3092/artifacts/, but there are no master logs. @kubernetes/goog-testing 

So i found another log:
https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/22263/kubernetes-pull-build-test-e2e-gce/33176/artifacts/e2e-gce-builder-4-2-minion-lwsg/?pli=1

```
I0321 19:31:10.613924       6 handlers.go:152] GET /api/v1/namespaces/e2e-tests-proxy-v7awn/pods/proxy-service-ak2qu-nfirt:80/proxy/: (2.588842ms) 503
goroutine 61736 [running]:
k8s.io/kubernetes/pkg/httplog.(*respLogger).recordStatus(0xc20af35110, 0x1f7)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/httplog/log.go:214 +0xa6
k8s.io/kubernetes/pkg/httplog.(*respLogger).WriteHeader(0xc20af35110, 0x1f7)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/httplog/log.go:193 +0x32
k8s.io/kubernetes/pkg/apiserver/metrics.(*responseWriterDelegator).WriteHeader(0xc20d1b1860, 0x1f7)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/metrics/metr
...
I0321 19:31:11.013029       6 handlers.go:152] GET /api/v1/nodes: (3.640252ms) 200 [[e2e.test/v1.3.0 (linux/amd64) kubernetes/18db078] 104.154.18.12:44594]
I0321 19:31:11.015321       6 handlers.go:152] GET /api/v1/namespaces/e2e-tests-proxy-v7awn/pods/http:proxy-service-ak2qu-nfirt:80/proxy/: (3.890299ms) 200 [[e2e.test/v1.3.0 (linux/amd64) kubernetes/18db078] 104.154.18.12:45954]
...
I0321 19:33:28.470836       6 handlers.go:152] DELETE /api/v1/namespaces/e2e-tests-proxy-v7awn/pods/proxy-service-ak2qu-nfirt: (12.634689ms) 200 [[kubelet/v1.3.0 (linux/amd64) kubernetes/18db078] 10.240.0.6:32979]
```

there was only 1 failure on the proxy url from apiserver logs and that was after the pod started. In kubelet logs:
```

I0321 19:31:09.717443    3398 prober.go:160] HTTP-Probe Headers: map[]
I0321 19:31:09.717553    3398 manager.go:190] Container readiness unchanged (false): "proxy-service-ak2qu-nfirt_e2e-tests-proxy-v7awn(719a751e-ef9b-11e5-a5b9-42010af00002)" - "docker://72d185d013202d2c05f37f90bb01c86a303580bc885ed92d54a00e68d7c6fc41"
I0321 19:31:09.718636    3398 http.go:79] Probe succeeded for http://10.245.5.7:80/, Response: {200 OK 200 HTTP/1.1 1 1 map[Date:[Mon, 21 Mar 2016 19:31:09 GMT] Content-Length:[29] Content-Type:[text/html; charset=utf-8]] 0xc208e7fe00 29 [] false map[] 0xc2087c7ee0 <nil>}
...
I0321 19:31:11.001565    3398 prober.go:111] Readiness probe for "proxy-service-ak2qu-nfirt_e2e-tests-proxy-v7awn(719a751e-ef9b-11e5-a5b9-42010af00002):proxy-service-ak2qu" succeeded
```

So maybe the test just needs to wait on readienss of an endpoint? Eitherway not sure this is team/cluster

**Comment by ixdy** [comment]
We don't have access to master logs on GKE.

**Comment by gmarek** [comment]
We don't have access to logs for real masters on GKE. We do have access to master logs in test projects. @roberthbailey @cjcullen

**Comment by ixdy** [comment]
The log dump scripts use ssh, and I don't think we have ssh access to masters anywhere.

**Comment by gmarek** [comment]
This is true;)

**Comment by a-robinson** [comment]
Googlers have access to the master logs from test projects, as @gmarek mentions. This certainly isn't as useful as if everyone did, but they're available when needed and access to them is documented internally.

**Comment by spxtr** [comment]
This test is exceptionally flaky on GKE (~6% flake rate). [Here](http://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/kubernetes-e2e-gke/6525) is a recent flake.

**Comment by gmarek** [comment]
This is a P0 since Mar 4th, so over 2 months now. Can we really, really, really prioritize this? @goltermann

**Comment by bprashanth** [comment]
I think the first problem is wrong triage. Who owns apiserver proxy? I debugged it simply because it was labelled team/cluster and figured it isn't something that worried me. In fact it seemed like a test fix  (https://github.com/kubernetes/kubernetes/issues/21484#issuecomment-213555886)

**Comment by gmarek** [comment]
@thockin - can you please reassign if cluster team is a wrong one? This is noticeably flaky.

**Comment by gmarek** [comment]
https://console.cloud.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gke/7303

**Comment by lavalamp** [comment]
https://console.cloud.google.com/storage/kubernetes-jenkins/logs/kubernetes-e2e-gke/7336/

**Comment by gmarek** [comment]
@lavalamp - should this be assigned to API machinery team?

**Comment by lavalamp** [comment]
@gmarek yes, probably :(

**Comment by sttts** [comment]
Some analysis:

- because the proxy error log output of https://github.com/golang/go/blob/91911e39f0fdfecc5453f9eca7ff74215ffb28a2/src/net/http/httputil/reverseproxy.go#L212 is not visible, the `WriteHeader` call we see in https://github.com/kubernetes/kubernetes/issues/21484#issuecomment-193437399 is actually this: https://github.com/golang/go/blob/91911e39f0fdfecc5453f9eca7ff74215ffb28a2/src/net/http/httputil/reverseproxy.go#L233
  => the 50x error actually originates from the proxy-service pod, not the apiserver
- a 50x from the proxy-service pod also means that it's not a readiness issue as @ixdy mentions in https://github.com/kubernetes/kubernetes/issues/21484#issuecomment-187451959 and also tries to fix with a merged PR
- it is always port 80 accessed via the pod url, not via the service name
- it is always the first access to the url
- it **does not** matter, whether other requests has been made to the pod-service pod before.

**Comment by caesarxuchao** [comment]
@sttts do you want to take this one? I haven't started working on it (I was OOO until this week) and have no prior knowledge.

**Comment by sttts** [comment]
@caesarxuchao @krousey I can continue to investigate that. Feel free to assign it to me.

**Comment by dims** [comment]
Is this a dup of the bot proposed one? https://github.com/kubernetes/kubernetes/issues/26210

**Comment by krousey** [comment]
Closing in favor of the automated report #26210
